ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -705,5 +733,19 @@ private <T> Callback<T> deleteDatasetVersionIfUploadFailedCallBack(Callback<T> c
         }
       };
     }
+
+    /**
+     * If updateVersionStateAndDeleteDatasetVersionOutOfRetentionCount failed, this method is to help set the rest method
+     * back to PUT.
+     * @param callback callback the final callback which submit the response.
+     */
+    private <T> Callback<T> setRestMethodBackToPutCallback(Callback<T> callback) {
+      return (r, e) -> {
+        restRequest.setRestMethod(RestMethod.PUT);
```

Comment:
can we create new implementation of RestRequest that takes in the `restRequest` when it needs to fetch headers, but return DELETE when it returns RestMethod. This way we would have a new object for delete handler so we don't have to reset the RestMethod.

Reply:
Thanks! Updated to use the WrappedRestRequest which always return DELETE restMethod.